api/src/main/java/uk/gov/hmcts/payment/api/reports/FeesService.java

```diff
@@ -0,0 +1,94 @@
+package uk.gov.hmcts.payment.api.reports;
+
+import org.slf4j.Logger;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import uk.gov.hmcts.fees2.register.api.contract.Fee2Dto;
+import uk.gov.hmcts.fees2.register.api.contract.FeeVersionDto;
+import uk.gov.hmcts.payment.api.contract.FeeDto;
+import uk.gov.hmcts.payment.api.contract.PaymentDto;
+
+import java.time.LocalDate;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+@Service
+@Transactional
+public class FeesService {
+    private static final Logger LOG = getLogger(FeesService.class);
+
+
+    private FeesRegisterClient feesRegisterClient;
+
+    private Map<String, Fee2Dto> feesDataMap = Collections.emptyMap();
+
+    private LocalDate feesDataMapRefreshedDate = LocalDate.now();
+
+    public FeesService(FeesRegisterClient feesRegisterClient) {
+        this.feesRegisterClient = feesRegisterClient;
+
+    }
+
+
+    public FeeVersionDto getFeeVersion(String feeCode, String feeVersion) {
+        Fee2Dto feeFromFeesRegister = getFeesDataMap().get(feeCode);
+        FeeVersionDto matchingFeeVersion = null;
+        if (null != feeFromFeesRegister) {
+            if (feeVersion.equals(feeFromFeesRegister.getCurrentVersion().getVersion().toString())) {
+                matchingFeeVersion = feeFromFeesRegister.getCurrentVersion();
+            } else {
+                Optional<FeeVersionDto> optionalMatchingFeeVersionDto = feeFromFeesRegister.getFeeVersionDtos()
+                    .stream().filter(versionDto -> versionDto.getVersion().equals(feeCode)).findFirst();
+                if (optionalMatchingFeeVersionDto.isPresent())
+                    matchingFeeVersion = optionalMatchingFeeVersionDto.get();
+            }
+        }
+        return matchingFeeVersion;
+    }
+
+    public List<PaymentDto> getMemolineAndNacForReconciliation(List<PaymentDto> payments) {
```

Comment:
remove this method and replace it wih a map lookup using key like feecode_version and FeeVersionDTO as value. i.e. prepare the map once retrieved from the fees register